src/uct/ib/ud/base/ud_ep.c

```diff
@@ -1281,13 +1301,47 @@ static uct_ud_send_skb_t *uct_ud_ep_prepare_crep(uct_ud_ep_t *ep)
     uct_ud_peer_name(ucs_unaligned_ptr(&crep->peer));
 
     skb->len = sizeof(*neth) + sizeof(*crep);
-    uct_ud_ep_ctl_op_del(ep, UCT_UD_EP_OP_CREP);
+
+    return skb;
+}
+
+static uct_ud_send_skb_t *uct_ud_ep_prepare_fin(uct_ud_ep_t *ep)
+{
+    uct_ud_iface_t *iface = ucs_derived_of(ep->super.super.iface,
+                                           uct_ud_iface_t);
+    uct_ud_send_skb_t *skb;
+    uct_ud_neth_t *neth;
+    uct_ud_ctl_hdr_t *fin;
+
+    skb = uct_ud_iface_get_tx_skb(iface, ep);
+    if (skb == NULL) {
+        return NULL;
+    }
+
+    neth = skb->neth;
+    uct_ud_neth_init_data(ep, neth);
+
+    neth->packet_type       = ep->dest_ep_id | UCT_UD_PACKET_FLAG_CTL;
+    fin                     = (uct_ud_ctl_hdr_t*)(neth + 1);
+    fin->type               = UCT_UD_PACKET_FIN;
+
+    uct_ud_peer_name(ucs_unaligned_ptr(&fin->peer));
+
+    skb->len = sizeof(*neth) + sizeof(*fin);
```

Comment:
can we unite common code with uct_ud_ep_prepare_crep?
the common function will initialize common neth fields and return pointer the the control packet (fin/crep). then, we fill the control packet fields in the separate functions, such as fin->type, crep->conn_rep, etc.